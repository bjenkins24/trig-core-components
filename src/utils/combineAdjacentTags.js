const combineAdjacentTags = ({ string, tag }) => {
  const splitOnOpen = string.split(`<${tag}>`);
  const finalSplit = splitOnOpen.map((newString) => {
    if (newString.includes(`</${tag}>`)) {
      return newString.split(`</${tag}>`);
    }
    return newString;
  });
  const finalArray = [];
  finalSplit.forEach((item, itemKey) => {
    if (typeof item === 'string') {
      return finalArray.push(item);
    }
    item.forEach((splitItem, splitItemKey) => {
      // This is the first item with a highlight that's been matched
      /* istanbul ignore else */
      if (
        (typeof finalSplit[itemKey - 1] === 'string' ||
          finalSplit[itemKey - 1][1] !== ' ') &&
        splitItemKey === 0
      ) {
        finalArray.push(`<${tag}>${splitItem}`);
      } else if (finalArray[finalArray.length - 1].includes(`<${tag}>`)) {
        finalArray[finalArray.length - 1] = `${
          finalArray[finalArray.length - 1]
        }${splitItem}`;
      }
      // If the second key is _not_ a space that means the first item is the last one
      if (item[1] !== ' ' && splitItemKey === 0) {
        finalArray[finalArray.length - 1] = `${
          finalArray[finalArray.length - 1]
        }</${tag}>`;
      }
    });
    return true;
  });

  return finalArray.reduce((accumulator, part) => {
    return `${accumulator}${part}`;
  }, '');
};

export { combineAdjacentTags };
