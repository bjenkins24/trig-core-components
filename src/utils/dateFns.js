import format from 'date-fns/format';
import subMonths from 'date-fns/subMonths';

export { format, subMonths };
