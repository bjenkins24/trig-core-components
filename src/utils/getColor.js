import { get } from 'lodash';

const getColor = (defaultColor = 'p') => ({ theme, color }) => {
  return get(theme, color, get(theme, defaultColor));
};

const getColorContrast = (defaultColor = 'p') => ({ theme, color }) => {
  return getColor(`${defaultColor}c`)({ theme, color: `color${color}` });
};

const getColorContrastShade = ({ defaultColor = 'p', shade = 600 }) => ({
  theme,
  color,
}) => {
  return getColorContrast(`${defaultColor}.${shade}`)({
    theme,
    color: `${color}.${shade}`,
  });
};

const getColorShade = ({ defaultColor = 'p', shade = 600 }) => ({
  theme,
  color,
}) => {
  return getColor(`${defaultColor}.${shade}`)({
    theme,
    color: `${color}.${shade}`,
  });
};

export { getColor, getColorShade, getColorContrast, getColorContrastShade };
