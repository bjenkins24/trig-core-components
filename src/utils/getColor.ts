import { get } from 'lodash';
import { Theme } from './theme';

interface ColorProps {
  theme: Theme;
  color: string;
}

const getColor = (defaultColor = 'p') => ({ theme, color }: ColorProps) => {
  if (color && (color.includes('#') || color.includes('rgb'))) {
    return color;
  }
  return get(theme, color, get(theme, defaultColor));
};

const getColorContrast = (defaultColor = 'p') => ({
  theme,
  color,
}: ColorProps) => {
  return getColor(`${defaultColor}c`)({ theme, color: `color${color}` });
};

const getColorContrastShade = ({ defaultColor = 'p', shade = 600 }) => ({
  theme,
  color,
}: ColorProps) => {
  return getColorContrast(`${defaultColor}.${shade}`)({
    theme,
    color: `${color}.${shade}`,
  });
};

const getColorShade = ({ defaultColor = 'p', shade = 600 }) => ({
  theme,
  color,
}: ColorProps) => {
  return getColor(`${defaultColor}.${shade}`)({
    theme,
    color: `${color}.${shade}`,
  });
};

export { getColor, getColorShade, getColorContrast, getColorContrastShade };
