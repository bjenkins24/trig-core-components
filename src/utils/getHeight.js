import { css } from 'styled-components';

const getHeight = ({ height }) => {
  /* istanbul ignore next */
  if (!height) return false;
  if (
    (typeof height === 'string' && height.includes('%')) ||
    height === 'auto'
  ) {
    return css`
      height: ${height};
    `;
  }

  return css`
    height: ${height}rem;
  `;
};

export default getHeight;
