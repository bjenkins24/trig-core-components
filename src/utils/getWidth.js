import { css } from 'styled-components';

const getWidth = ({ width }) => {
  if (typeof width === 'string' && width.includes('%')) {
    return css`
      width: ${width};
    `;
  }

  return css`
    width: ${width}rem;
  `;
};

export default getWidth;
