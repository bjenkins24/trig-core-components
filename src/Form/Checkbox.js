import React, { useState, useRef } from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'styled-components';
import { uniqueId } from 'lodash';
import { widthType } from 'utils/propTypes';
import { HorizontalGroup } from 'Groups';
import ErrorMessage from './ErrorMessage';
import Label from './Label';

import Icon from '../Icon';

const LabelContainer = styled(Label).attrs({
  'data-testid': 'labelContainer',
})`
  padding-left: ${({ labelPosition }) =>
    labelPosition === 'right' ? '1.6rem' : 0};
`;

const checkedStyles = css`
  border: 0.1rem solid ${({ theme }) => theme.s};
  background: ${({ theme }) => theme.s};
`;

const uncheckedStyles = css`
  border: 0.1rem solid ${({ theme }) => theme.ps[100]};
  background: none;
`;

const StyledCheckbox = styled.div`
  display: flex;
  align-self: flex-start;
  margin-top: 0.4rem;
  flex-shrink: 0;
  width: 1.6rem;
  height: 1.6rem;
  cursor: pointer;
  color: ${({ theme }) => theme.bs[200]};
  border-radius: 0.2rem;
  ${uncheckedStyles}
  &:hover {
    border: 0.1rem solid ${({ theme }) => theme.s};
  }
`;

const HiddenInput = styled.input.attrs({
  type: 'checkbox',
})`
  /* Source: https://polished.js.org/docs/#hidevisually */
  border: 0;
  height: 1px;
  margin: -1px;
  clip: rect(0 0 0 0);
  clip-path: inset(50%);
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;

  :checked + div {
    ${checkedStyles}
  }
  &:not(:checked):focus ~ ${StyledCheckbox} {
    border: 0.1rem solid ${({ theme }) => theme.s};
  }
  &:checked:focus ~ ${StyledCheckbox} {
    border: 0.1rem solid ${({ theme }) => theme.ss[800]};
  }
`;

const StyledIcon = styled(Icon)`
  margin: 0 auto;
  align-self: center;
`;

export const checkboxTypes = {
  label: PropTypes.oneOfType([PropTypes.func, PropTypes.string]),
  children: PropTypes.func,
  className: PropTypes.string,
  checked: PropTypes.bool,
  onChange: PropTypes.func,
  width: widthType,
  labelPosition: PropTypes.oneOf(['left', 'right']),
  labelProps: PropTypes.object,
  error: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.string,
    PropTypes.bool,
  ]),
};

const defaultProps = {
  label: '',
  labelProps: {},
  children: null,
  className: '',
  checked: false,
  onChange: () => null,
  width: null,
  labelPosition: 'left',
  error: null,
};

const Checkbox = ({
  label,
  labelProps,
  labelPosition,
  children,
  className,
  checked,
  onChange,
  width,
  error,
  ...restProps
}) => {
  const [forId] = useState(() => uniqueId('checkbox-'));
  const inputRef = useRef(null);
  const [isChecked, setIsChecked] = useState(checked);

  const renderLabel = () => {
    return (
      <LabelContainer
        width={width}
        labelPosition={labelPosition}
        color="ps.200"
        {...labelProps}
        htmlFor={forId}
      >
        {label}
      </LabelContainer>
    );
  };

  return (
    <div className={className}>
      <HorizontalGroup padding={labelPosition === 'right' ? 0 : 1.6}>
        {label && labelPosition === 'left' && renderLabel()}
        <HiddenInput
          ref={inputRef}
          id={forId}
          checked={isChecked}
          onChange={(e) => {
            setIsChecked(!isChecked);
            onChange(e);
          }}
          {...restProps}
        />
        <StyledCheckbox
          data-testid="checkbox-icon"
          onClick={() => {
            inputRef.current.click();
          }}
        >
          {isChecked && <StyledIcon type="check" size={1.2} />}
        </StyledCheckbox>
        {label && labelPosition === 'right' && renderLabel()}
      </HorizontalGroup>
      {error && <ErrorMessage>{error}</ErrorMessage>}
    </div>
  );
};

Checkbox.propTypes = checkboxTypes;
Checkbox.defaultProps = defaultProps;

export default Checkbox;
