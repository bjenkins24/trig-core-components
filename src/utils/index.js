import {
  getColor,
  getColorShade,
  getColorContrast,
  getColorContrastShade,
} from './getColor';
import getWidth from './getWidth';
import getHeight from './getHeight';
import useTheme from './useTheme';

export {
  getColor,
  getColorShade,
  getColorContrast,
  getColorContrastShade,
  getWidth,
  getHeight,
  useTheme,
};
