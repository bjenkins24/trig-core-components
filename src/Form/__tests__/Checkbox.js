import React from 'react';
import { render } from 'test/utils';
import user from '@testing-library/user-event';
import Checkbox from 'Form/Checkbox';

const labelText = 'My Checkbox';

describe('<Checkbox />', () => {
  it('renders and takes basic props', () => {
    const { getByText, getByRole, getByTestId, rerender } = render(
      <Checkbox />
    );

    expect(getByTestId('labelContainer')).toBeInTheDocument();
    expect(getByRole('checkbox')).toBeInTheDocument();

    rerender(<Checkbox label={labelText} />);
    expect(getByText(labelText)).toBeInTheDocument();
  });

  it('renders custom with render props', () => {
    const randomText = 'Custom Label';
    const { getByRole, getByText, getByTestId } = render(
      <Checkbox>
        {({ renderCheckbox }) => {
          return (
            <div>
              {randomText} {renderCheckbox()}
            </div>
          );
        }}
      </Checkbox>
    );

    // It needs to be in a label container
    expect(getByTestId('labelContainer')).toBeInTheDocument();
    expect(getByRole('checkbox')).toBeInTheDocument();
    expect(getByText(randomText)).toBeInTheDocument();
  });

  it('shows and hides check', () => {
    const { getByTitle, getByRole, queryByTitle } = render(
      <Checkbox checked />
    );
    expect(getByTitle(/check icon/i)).toBeInTheDocument();
    user.click(getByRole('checkbox'));
    expect(queryByTitle(/check icon/i)).toBeNull();
  });

  it('takes custom onChange', () => {
    const mockCallback = jest.fn();
    const { getByRole } = render(<Checkbox onChange={mockCallback} />);
    user.click(getByRole('checkbox'));
    expect(mockCallback.mock.calls.length).toEqual(1);
  });
});
