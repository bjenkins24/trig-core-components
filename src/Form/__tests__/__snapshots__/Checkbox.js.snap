// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Checkbox /> shows and hides check 1`] = `
.c7 {
  font-size: 1.6rem;
  line-height: 1.7;
  font-weight: 400;
  color: #8A8787;
}

.c8 {
  cursor: pointer;
  width: 20rem;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
}

.c6 {
  color: inherit;
  margin: 0 auto;
  -webkit-transition: all 0.15s;
  transition: all 0.15s;
}

.c9 {
  padding-left: 0.8rem;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
  margin-top: 0.4rem;
  width: 1.6rem;
  height: 1.6rem;
  cursor: pointer;
  color: #FFFFFF;
  border-radius: 0.2rem;
  border: 0.1rem solid #C0BFBF;
  background: none;
}

.c3:hover {
  border: 0.1rem solid #009E8F;
}

.c1 {
  border: 0;
  height: 1px;
  margin: -1px;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.c1:checked + div {
  border: 0.1rem solid #009E8F;
  background: #009E8F;
}

.c1:not(:checked):focus ~ .c2 {
  border: 0.1rem solid #009E8F;
}

.c1:checked:focus ~ .c2 {
  border: 0.1rem solid #098072;
}

.c5 {
  margin: 0 auto;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
}

<div
  class=""
>
  <div
    class="c0"
  >
    <input
      checked=""
      class="c1"
      id="checkbox-2"
      type="checkbox"
    />
    <div
      class="c2 c3"
      data-testid="checkbox-icon"
    >
      <div
        class="c4 c5"
      >
        <svg
          class="c6"
          height="1.2rem"
          role="img"
          viewBox="0 0 32 32"
          width="1.2rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M29.543 4L9.918 23.193l-7.461-7.296L0 18.3 9.918 28 32 6.403z"
            fill="currentColor"
            fill-rule="evenodd"
          />
          <title>
            Check Icon
          </title>
          <desc />
        </svg>
      </div>
    </div>
    <label
      class="c7 c8 c9"
      color="ps.200"
      data-testid="labelContainer"
      for="checkbox-2"
      width="20"
    >
      hello
    </label>
  </div>
</div>
`;
